src/Functional/ChunkEvery.php

```diff
@@ -0,0 +1,64 @@
+<?php
+/**
+ * Copyright (C) 2019 by Bas Bloembergen <basbloembergen@gmail.com>
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+namespace Functional;
+
+use Functional\Exceptions\InvalidArgumentException;
+use Traversable;
+
+/**
+ * @param Traversable|array $collection
+ * @param int $size Desired size of the chunk
+ * @param int $step Step with which to increment start of next chunk
+ * @param bool $discardIncompleteChunk Discard the last chunk if it does not have the specified size.
+ * @return array
+ */
+function chunk_every($collection, $size = 1, $step = 1, $discardIncompleteChunk = false)
+{
+    InvalidArgumentException::assertCollection($collection, __FUNCTION__, 1);
+
+    if ($collection instanceof Traversable) {
+        $array = \iterator_to_array($collection);
```

Comment:
This won’t work with an infinite iterator or an iterator where the keys are duplicated

Reply:
You're right, I'll tinker some more!

Reply:
What kind of solution would you prefer for the InfiniteIterator case? Is reducing it to its inner iterator acceptable?